contracts/TwabRewards.sol

```diff
@@ -0,0 +1,394 @@
+            _startTimestamp,
+            _epochDuration,
+            _numberOfEpochs
+        );
+
+        _promotions[_nextPromotionId] = _nextPromotion;
+
+        emit PromotionCreated(_nextPromotionId);
+
+        return _nextPromotionId;
+    }
+
+    /// @inheritdoc ITwabRewards
+    function cancelPromotion(uint256 _promotionId, address _to)
+        external
+        override
+        onlyPromotionCreator(_promotionId)
+        returns (bool)
+    {
+        require(_isPromotionActive(_promotionId) == true, "TwabRewards/promotion-not-active");
+        require(_to != address(0), "TwabRewards/recipient-not-zero-address");
+
+        Promotion memory _promotion = _getPromotion(_promotionId);
+
+        IERC20 _token = IERC20(_promotion.token);
+        uint256 _remainingRewards = _getRemainingRewards(_promotionId);
+
+        if (_remainingRewards > 0) {
+            _token.safeTransfer(_to, _remainingRewards);
+        }
+
+        delete _promotions[_promotionId];
+
+        emit PromotionCancelled(_token, _remainingRewards);
+
+        return true;
+    }
+
+    /// @inheritdoc ITwabRewards
+    function extendPromotion(uint256 _promotionId, uint8 _numberOfEpochs)
+        external
+        override
+        returns (bool)
+    {
+        require(_isPromotionActive(_promotionId) == true, "TwabRewards/promotion-not-active");
+
+        Promotion memory _promotion = _getPromotion(_promotionId);
+        uint8 _extendedNumberOfEpochs = _promotion.numberOfEpochs + _numberOfEpochs;
+
+        _requireEpochLimit(_extendedNumberOfEpochs);
+
+        IERC20 _token = IERC20(_promotion.token);
+        uint256 _amount = _numberOfEpochs * _promotion.tokensPerEpoch;
+
+        _token.safeTransfer(address(this), _amount);
+        _promotions[_promotionId].numberOfEpochs = _extendedNumberOfEpochs;
+
+        emit PromotionExtended(_token, _amount, _extendedNumberOfEpochs);
+
+        return true;
+    }
+
+    /// @inheritdoc ITwabRewards
+    function getPromotion(uint256 _promotionId) external view override returns (Promotion memory) {
+        return _getPromotion(_promotionId);
+    }
+
+    /// @inheritdoc ITwabRewards
+    function getRemainingRewards(uint256 _promotionId) external view override returns (uint256) {
+        return _getRemainingRewards(_promotionId);
+    }
+
+    /// @inheritdoc ITwabRewards
+    function getCurrentEpochId(uint256 _promotionId) external view override returns (uint256) {
+        return _getCurrentEpochId(_promotionId);
+    }
+
+    /// @inheritdoc ITwabRewards
+    function getRewardAmount(
+        address _user,
+        uint256 _promotionId,
+        uint256 _epochId
+    ) external view override returns (uint256) {
+        return _calculateRewardAmount(_user, _promotionId, _epochId);
+    }
+
+    /// @inheritdoc ITwabRewards
+    function claimRewards(
+        address _user,
+        uint256 _promotionId,
+        uint256 _epochId
+    ) external override returns (uint256) {
```

Comment:
still waiting for batch!

Reply:
Done: https://github.com/pooltogether/v4-periphery/pull/8/commits/bc841db2c168e6177abba66d47189686e1a6aa79